### PR TITLE
Define sanitization for db.query.text

### LINKED ISSUE #717: DB sanitization uniform format
**What are you trying to achieve?**

According to the [db spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#call-level-attributes), the `db.statement` value can be sanitized, but it is not defined how to do so.

Currently, the sanitization is being dealt with differently in few places.
I suggest to add a uniform format that will describe how to do the sanitization.
(It will be best if this format will apply to all different DB's and syntaxes)

Different implementations examples:
- [JS mongo db](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-mongodb/src/instrumentation.ts#L606) - implements the sanitization by replacing the information with question marks
- [Python pymongo](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py#L202) - implements the sanitization by deleting the information completely, and leaving the query method name only 
- [Python elasticsearch (WIP) ](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/1598) - suggests to replace the information with a string that will explain that the data is sanitized

I suggest a few options to replace the value with:

- **Keep the method name, and add a sanitized text.** 
**for example:** `db.statement = "SELECT {query information is sanitized}"`
**advantages:** quite easy to implement, easy to keep consistent across libraries.
**dis-advantages:** will require to research whether all different libraries can handle this format effectively.

- **Simple text that will describe that the value is sanitized.**
**for example:** `db.statement = "query information is sanitized"`
**advantages:** easy to implement, easy to keep consistent across different libraries.
**dis-advantages:** doesn't supply basic information about the query that could be useful.

- **Replace the values with question marks.**
 **for example:** `db.statement = "SELECT ? FROM ?"`
 **advantages:** keeps more amount of information, while still not exposing sensitive or private data.
 **dis-advantages:** harder to implement, harder to keep consistent across libraries. 
 
 I would like to hear opinions about the suggested solutions, or hear different ideas.

**Additional context.**

open-telemetry/opentelemetry-specification#3104 - Issue regarding changing the recommendation to sanitize the information by default.
open-telemetry/semantic-conventions#708 - Issue about missing examples for sanitization in specs.

### PR DESCRIPTION
**Pull request by gregkalapos**
Fixes #717

## Changes

Describes how to sanitize `db.query.text`. 

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [ ] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

File: model/trace/database.yaml [hunk 1]
```diff
@@ -5,18 +5,7 @@ groups:
...
-            Non-parameterized query text SHOULD NOT be collected by default unless there is sanitization that excludes
-            sensitive data, e.g. by redacting all literal values present in the query text.
-
-            Parameterized query text SHOULD be collected by default
-            (the query parameter values themselves are opt-in,
-            see [`db.query.parameter.<key>`](../../docs/attributes-registry/db.md)).
-        note:
-          Even though parameterized query text can potentially have sensitive data, by using a parameterized query
-          the user is giving a strong signal that any sensitive data will be passed as parameter values, and the benefit
-          to observability of capturing the static part of the query text by default outweighs the risk.
```
Comment (gregkalapos) [inline model/trace/database.yaml:19]: I'm not sure this was intentionally defined here in `attributes.db.client.minimal`. This was overwritten below in `trace.db.common.query_and_collection` (which extends `attributes.db.client.minimal`), which resulted in this text never being rendered in any of the `.md` files. Some specific db specs extend `attributes.db.client.minimal`, but none of them seem to use `db.query.text`.

So I moved the whole text below to `trace.db.common.query_and_collection`, which seem to make more sense (e.g. that way `database-spans.md` is correct, but may not be ok for some other specs).

I think this should be cleaned up - let's discuss in the WG meeting.

File: docs/database/cosmosdb.md [hunk 2]
```diff
@@ -61,9 +61,10 @@ For batch operations, if the individual operations are known to have the same op
 
 **[7]:** If using a port other than the default port for this DBMS and if `server.address` is set.
 
-**[8]:** For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.
+**[8]:** Even though parameterized query text can potentially have sensitive data, by using a parameterized query the user is giving a strong signal that any sensitive data will be passed as parameter values, and the benefit to observability of capturing the static part of the query text by default outweighs the risk.
```
Comment (lmolkova) [inline docs/database/cosmosdb.md:64]: this change seems off. We used to document what it should be be batch operations and now we talk about parametrization.

Perhaps we want both parts of this note to be present (here and in the registry)?

I have no objections addressing it here or moving to a separate PR.

File: docs/database/cosmosdb.md [hunk 3]
```diff
@@ -61,9 +61,10 @@ For batch operations, if the individual operations are known to have the same op
...
 
-**[9]:** SHOULD be collected by default only if there is sanitization that excludes sensitive information.
+**[9]:** Non-parameterized query text SHOULD NOT be collected by default unless there is sanitization that excludes sensitive data. Sanitization SHOULD be done by replacing all literals with a placeholder value. Such literals include, but are not limited to, String literals, Numeric literals, Date and Time literals, Boolean literals, Interval literals, Binary literals, and Hexadecimal literals. The placeholder value SHOULD be `?`, unless it already has a defined meaning in the given database system, in which case the instrumentation MAY choose a different placeholder.
```
Comment (lmolkova) [inline docs/database/cosmosdb.md:66]: I think this text belongs on the attribute note (and not on the requirement level) since it talks about the format, not the condition under which the attribute should be collected. 
Also examples in the second sentence are definitely related to the existing parametrization note.

Comment (maryliag) [reply docs/database/cosmosdb.md:66]: I wonder if the rules/format of how to sanitize should be it's own section (or maybe file), since can be complex. This way we can add examples, considerations, etc, and then all places can just point to that link.

Then I can add the rules of grouping variables (issue #1053) on the same place.

Comment (gregkalapos) [reply docs/database/cosmosdb.md:66]: Yeah, totally agree, we should move this.

> should be it's own section (or maybe file)

So where should we move it? The big advantage of having this in yaml is that it'll be always auto generated (like this md file) - the downside is that the best we have there is to put this into a `note`, right? 

If we move it into a file, then we'll need to manually maintain it.

What should we do?

Comment (matt-hensley) [reply docs/database/cosmosdb.md:66]: From June 5 SIG call: Is there any way to determine whether a query has been sanitized? `?` is a common placeholder for query parameters.

An explicit flag that a query was sanitized could work, with the benefit of not require any string parsing to determine if the query was originally parameterized or not.

Comment (maryliag) [reply docs/database/cosmosdb.md:66]: > If we move it into a file, then we'll need to manually maintain it.

I see this as the equivalent of manually maintaining the yaml file. If there is a new md file, your yaml could have the note pointing to it, so that part continues to be auto generated.

The new file could be on `docs/database/sanitization.md`

File: model/trace/database.yaml (hunk 1 above)
Comment (gregkalapos) [reply model/trace/database.yaml]: Should be ok now.

File: docs/database/cosmosdb.md (hunk 3 above)
Comment (gregkalapos) [reply docs/database/cosmosdb.md]: Discussed in the WG meeting - we add it to `database-spans.md` for now. Happy to move it out later if needed. We link to that section from yaml.

Comment (maryliag) [reply docs/database/cosmosdb.md]: Awesome, I like how it looks now! Thank you for making the update!

**Review by maryliag — APPROVED** [review]

File: docs/database/database-spans.md [hunk 4]
```diff
@@ -76,6 +77,17 @@ and SHOULD adhere to one of the following values, provided they are accessible:
 If a corresponding `{target}` value is not available for a specific operation, the instrumentation SHOULD omit the `{target}`.
 For example, for an operation describing SQL query on an anonymous table like `SELECT * FROM (SELECT * FROM table) t`, span name should be `SELECT`.
 
+## Sanitization of `db.query.text`
```
Comment (lmolkova) [inline docs/database/database-spans.md:80]: Can it be moved down to be under the attribute table? Then we'll first introduce the attribute and then talk about its sanitization.

File: docs/database/database-spans.md [hunk 5]
```diff
@@ -76,6 +77,17 @@ and SHOULD adhere to one of the following values, provided they are accessible:
...
+
+Sanitization MUST be done by replacing all literals with a placeholder value. Such literals
```
Comment (lmolkova) [inline docs/database/database-spans.md:82]: It could be interpreted as that sanitization is a MUST and I hope we can clarify it with something like

```suggestion
The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.

Sanitization SHOULD replace all literals with a placeholder value. Such literals
```

I intentionally changed MUST to SHOULD since the sanitization so far meant removing sensitive information, not all literals. Happy to discuss more.

But also it feels that we use MUST to tell **how** to do sanitization and it's too strong.

File: docs/database/database-spans.md [hunk 6]
```diff
@@ -125,7 +137,8 @@ For batch operations, if the individual operations are known to have the same op
 
 **[9]:** If using a port other than the default port for this DBMS and if `server.address` is set.
 
-**[10]:** For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.
+**[10]:** For sanitization see [Sanitization of `db.query.text`](../../docs/database/database-spans.md#sanitization-of-dbquerytext)
+For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.
 
 **[11]:** SHOULD be collected by default only if there is sanitization that excludes sensitive information.
```
Comment (lmolkova) [inline docs/database/database-spans.md:143]: let's reference the new section here instead of the current text in the note?

File: docs/database/database-spans.md (hunk 5 above)
Comment (gregkalapos) [reply docs/database/database-spans.md]: `SHOULD` sounds ok to me here as well. I was going back and forth between `SHOULD` and `MUST` - so let's just move on with `SHOULD`.

File: docs/database/database-spans.md (hunk 4 above)
Comment (gregkalapos) [reply docs/database/database-spans.md]: Makes sense, moved this part below.

File: docs/database/cosmosdb.md (hunk 3 above)
Comment (gregkalapos) [reply docs/database/cosmosdb.md]: > From June 5 SIG call: Is there any way to determine whether a query has been sanitized? `?` is a common placeholder for query parameters.
> 
> An explicit flag that a query was sanitized could work, with the benefit of not require any string parsing to determine if the query was originally parameterized or not.

That's a good point, however I'm not sure if this is important enough to address it at this point. We can always add an additional attribute to store if the query was sanitized or not as an incremental change.

There is the `db.query.parameter.<key>` to store parameters - so if the query is parametrized, then that attributes is used - however that attribute is opt-in, so admittedly that'll not help in a default setup.

I personally would wait for user feedback on this and if needed, then I'd add the additional flag on-demand if we really see the need for it.

File: docs/database/database-spans.md [hunk 7]
```diff
@@ -225,6 +227,18 @@ Back ends could, for example, use the provided identifier to determine the appro
...
+## Sanitization of `db.query.text`
+
+The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
+Sanitization SHOULD replace all literals with a placeholder value.
+Such literals include, but are not limited to, String literals, Numeric literals, Date and Time literals,
+Boolean literals, Interval literals, Binary literals, and Hexadecimal literals.
+The placeholder value SHOULD be `?`, unless it already has a defined meaning in the given database system,
+in which case the instrumentation MAY choose a different placeholder.
+
+Placeholders in a parametrized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,
```
Comment (trask) [inline docs/database/database-spans.md:239]:
```suggestion
Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,
```

File: docs/database/database-spans.md [hunk 8]
```diff
@@ -225,6 +227,18 @@ Back ends could, for example, use the provided identifier to determine the appro
...
+and there is no need to sanitize the value `$1`.
```
Comment (trask) [inline docs/database/database-spans.md:240]:
```suggestion
and there is no need to sanitize the text `$1`.
```

File: model/trace/database.yaml [hunk 9]
```diff
@@ -26,9 +28,15 @@ groups:
...
+        note: >
+          For sanitization see [Sanitization of `db.query.text`](../../docs/database/database-spans.md#sanitization-of-dbquerytext)
+
+          For batch operations, if the individual operations are known to have the same query text
+          then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated
+          with separator `; ` or some other database system specific separator if more applicable.
         requirement_level:
           recommended: >
-            SHOULD be collected by default only if there is sanitization that excludes sensitive information.
+            See [Sanitization of `db.query.text`](../../docs/database/database-spans.md#sanitization-of-dbquerytext).
```
Comment (trask) [inline model/trace/database.yaml:39]: what happens if we inherit note and requirement_level from `attributes.db.client.minimal`?

Comment (gregkalapos) [reply model/trace/database.yaml:39]: In short: for note, we'd loose the part that talks about batch operations, but for `requirement_level`, we could probably do that. A few things inherit from `trace.db.common.query_and_collection` (e.g. cassandra, sql, cosmos), so the effect of doing it for `requirement_level` will be that all those will have a longer text for `requirement_level` talking about parameterized and Non-parameterized queries. But I think that's fine.

Regarding `note`: there is also a note in the registry, which we seem to overwrite everywhere. I don't really understand yet why. Wouldn't it be easier to just handle the note for db.query.text in the registry and just inherit that everywhere? So far I don't see the logic of differentiating that note for the different ids.

File: docs/database/cassandra.md [hunk 10]
```diff
@@ -60,15 +60,15 @@ For batch operations, if the individual operations are known to have the same op
...
 
-**[9]:** For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.
+**[9]:** For sanitization see [Sanitization of `db.query.text`](../../docs/database/database-spans.md#sanitization-of-dbquerytext).
+For batch operations, if the individual operations are known to have the same query text then that query text SHOULD be used, otherwise all of the individual query texts SHOULD be concatenated with separator `; ` or some other database system specific separator if more applicable.
+Even though parameterized query text can potentially have sensitive data, by using a parameterized query the user is giving a strong signal that any sensitive data will be passed as parameter values, and the benefit to observability of capturing the static part of the query text by default outweighs the risk.
 
-**[10]:** SHOULD be collected by default only if there is sanitization that excludes sensitive information.
```
Comment (gregkalapos) [inline docs/database/cassandra.md:65]: This was a note on the `requirement_level`, which got removed here. We have a general note now for the attribute which starts with `For sanitization see Sanitization of db.query.text...`, which starts with the exact same sentence. Having this twice is not needed -> so this is dropped in other md files below as well.

File: model/trace/database.yaml (hunk 9 above)
Comment (gregkalapos) [reply model/trace/database.yaml]: I moved the general note to the registry and consolidated it there - I really wonder why it was overwritten [here](https://github.com/open-telemetry/semantic-conventions/blob/main/model/trace/database.yaml#L16-L19). 

With this, we just inherit the note from the registry which now covers the link to sanitization, batching, and parameterized query.

Let me know if this is not an ok move.

File: model/trace/database.yaml [hunk 11]
```diff
@@ -26,9 +22,6 @@ groups:
     brief: This group defines the attributes used to perform database client calls.
     attributes:
       - ref: db.query.text
-        requirement_level:
-          recommended: >
-            SHOULD be collected by default only if there is sanitization that excludes sensitive information.
```
Comment (lmolkova) [inline model/trace/database.yaml:31]: I feel it's important to preserve.

I understand it's not affecting markdown much, but at least in the yaml we specify this under requirement-level. 

Also we can always do better job changing how markdown is rendered and this may become significant there too.

File: docs/database/database-spans.md [hunk 12]
```diff
@@ -225,6 +226,18 @@ Back ends could, for example, use the provided identifier to determine the appro
 
 When additional attributes are added that only apply to a specific DBMS, its identifier SHOULD be used as a namespace in the attribute key as for the attributes in the sections below.
 
+## Sanitization of `db.query.text`
+
+The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
+Sanitization SHOULD replace all literals with a placeholder value.
+Such literals include, but are not limited to, String literals, Numeric literals, Date and Time literals,
+Boolean literals, Interval literals, Binary literals, and Hexadecimal literals.
```
Comment (lmolkova) [inline docs/database/database-spans.md:236]: super-nit: 

maybe
```suggestion
Such literals include, but are not limited to, String, Numeric, Date and Time,
Boolean, Interval, Binary, and Hexadecimal literals.
```

File: model/trace/database.yaml (hunk 11 above)
Comment (gregkalapos) [reply model/trace/database.yaml:31]: Ok, added it back.

File: docs/database/cassandra.md (hunk 10 above)
Comment (gregkalapos) [reply docs/database/cassandra.md:65]: Discussed in https://github.com/open-telemetry/semantic-conventions/pull/1100#discussion_r1649060948 - we add it back.

File: docs/database/database-spans.md [hunk 13]
```diff
@@ -225,6 +226,18 @@ Back ends could, for example, use the provided identifier to determine the appro
...
+The placeholder value SHOULD be `?`, unless it already has a defined meaning in the given database system,
+in which case the instrumentation MAY choose a different placeholder.
+
+Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is,
+and there is no need to sanitize the text `$1`.
```
Comment (lmolkova) [inline docs/database/database-spans.md:241]: nit:

```suggestion
Placeholders in a parameterized query SHOULD not be sanitized. E.g. `where id = $1` can be captured as is.
```

File: model/trace/database.yaml (hunk 11 above)
Comment (maryliag) [reply model/trace/database.yaml:31]: if is only required with condition, shouldn't this be using something like
```
 requirement_level:
          conditionally_required: If there is sanitization that excludes sensitive information.
```

File: docs/database/database-spans.md [hunk 14]
```diff
@@ -225,6 +228,18 @@ Back ends could, for example, use the provided identifier to determine the appro
 
 When additional attributes are added that only apply to a specific DBMS, its identifier SHOULD be used as a namespace in the attribute key as for the attributes in the sections below.
 
+## Sanitization of `db.query.text`
+
+The `db.query.text` SHOULD be collected by default only if there is sanitization that excludes sensitive information.
+Sanitization SHOULD replace all literals with a placeholder value.
+Such literals include, but are not limited to, String, Numeric, Date and Time,
+Boolean, Interval, Binary, and Hexadecimal literals.
```
Comment (maryliag) [inline docs/database/database-spans.md:236]: nit: you can probably remove the "literals" on the "Hexadecimal" too, since it already starts the sentence saying those are type of literals (I saw you removed from the other on your latest commit)

File: model/trace/database.yaml (hunk 11 above)
Comment (gregkalapos) [reply model/trace/database.yaml:31]: That'd be a stronger requirement level - it'd mean, if it's sanitized, then it must be captured. Right?

Now we just say, it's recommended, so it may not be captured and I can imagine there will be configs to opt in and out to capture this. So I feel recommended is appropriate here. That was already the case prior to the PR - so there would be no change on that here.

Comment (lmolkova) [reply model/trace/database.yaml:31]: The PR would not change the status quo, so I don't want to block it on the req_level discussion, but I agree that

> conditionally_required: If there is sanitization that excludes sensitive information.

would be more precise - let's chat on the SIG call and maybe change it in a follow up

File: docs/database/database-spans.md (hunk 14 above)
Comment (gregkalapos) [reply docs/database/database-spans.md:236]: I'm not a native speak, so I just asked AI and it says it's ok to drop it, but keeping it there is more precise. It says  `"literals" at the end makes it explicitly clear that all the items in the list are types of literals`. That's also how I'd understand it - the "literals" is not on the "Hexadecimal", but on the whole list.

File: docs/database/cosmosdb.md (hunk 3 above)
Comment (trask) [reply docs/database/cosmosdb.md]: > I personally would wait for user feedback on this and if needed, then I'd add the additional flag on-demand if we really see the need for it.

let's create an issue for this 👍

**Review by trask — APPROVED** [review]

**Review by lmolkova — APPROVED** [review]

File: docs/database/cosmosdb.md (hunk 3 above)
Comment (lmolkova) [reply docs/database/cosmosdb.md]: Created https://github.com/open-telemetry/semantic-conventions/issues/1179
